### DISCUSSION TITLE
useRecoilState with default value like useState

### DISCUSSION THREAD

**Discussion by iSplasher** [facebookexperimental/Recoil#1122] [Ideas]
Hello,
Is there any reason why the `useRecoilState` hook doesn't take an optional default value like `useState` does?
This would solve a lot of issues such as making data available on initial render and also hydrating data from the server.
In effect this would work like you had set the state with `useSetRecoilState` in a hook, but the benefit would be the value being available on initial render.
Is there any issue that I'm not seeing?

```typescript

function Component(){
const [state, setState] = useRecoilState(SomeState, "hello world")

// "hello world" is immediately on first render
// other components that depends on the value are naturally updated as well

return <div>{value}</div>
}
```

**Comment by iSplasher** [comment]
Something like this maybe?
```typescript
export function useInitialRecoilState<T>(state: RecoilState<T>, optional?: T) {
  const [value, setValue] = useRecoilState(state);
  const [optionalValue, setOptionalValue] = useState([value, optional]);

  useEffectOnce(() => {
    if (optional === value) {
      setOptionalValue(undefined);
    } else {
      setValue(optional);
    }
  });

  useUpdateEffect(() => {
    setOptionalValue(undefined);
  }, [value]);

  return [
    optionalValue === undefined || value !== optionalValue[0]
      ? value
      : optionalValue[1] ,
    setValue,
  ];
}
```

**Comment by drarmstr** [comment]
The default value for some atom state is managed by the `default` property when the atom itself is defined:
```jsx
const someState = atom({
  key: 'MyModule/SomeState',
  default: 'hello world',
});

function Component(){
  const [state, setState] = useRecoilState(someState);

  // "hello world" is immediately on first render
  return <div>{state}</div>;
}
```

**Reply by iSplasher** [reply]
Hello, thanks for the reply.
I am already aware of that.
I was proposing an improvement to the API.
In this case I'm trying to make state that I am getting from elsewhere, like the server, available immediately just like `initializeState` of `RecoilRoot` allows but the latter lacks flexibility (like not being compatible with atomFamily).
So I'm proposing an improvement to `useRecoilState` that takes an optional value that performs the same role.

**Reply by drarmstr** [reply]
Having a hook set the initial state that other components see is problematic as setting state is an effect wouldn't be part of rendering and depends on scheduling ordering of the different components and effects.

For syncing initial state with an external source consider [Atom Effects](https://recoiljs.org/docs/guides/atom-effects)

**Reply by iSplasher** [reply]
I see that makes sense, thanks!
